oak_io/src/handle.rs

```diff
@@ -59,6 +61,67 @@ impl From<Handle> for WriteHandle {
     }
 }
 
+/// Provides implementations of `Clone` and `Drop` for handle types, based on the `handle_clone` and
+/// `channel_close` ABI syscalls, respectively.
+#[cfg(feature = "linear-handles")]
+mod linear_handles {
+    use super::*;
+    use crate::OakStatus;
+    use log::trace;
+
+    /// Creates a new distinct handle using `oak_abi::clone_handle`
+    impl Clone for ReadHandle {
+        fn clone(&self) -> Self {
+            Self {
+                handle: clone_handle(self.handle),
+            }
+        }
+    }
+
+    /// Creates a new distinct handle using `oak_abi::clone_handle`
+    impl Clone for WriteHandle {
+        fn clone(&self) -> Self {
+            Self {
+                handle: clone_handle(self.handle),
+            }
+        }
+    }
+
+    /// Closes the handle using `oak_abi::channel_close`
+    impl Drop for ReadHandle {
+        fn drop(&mut self) {
+            drop_handle(self.handle);
+        }
+    }
+
+    /// Closes the handle using `oak_abi::channel_close`
+    impl Drop for WriteHandle {
+        fn drop(&mut self) {
+            drop_handle(self.handle);
+        }
+    }
+
+    fn clone_handle(handle: Handle) -> Handle {
+        let mut cloned_handle = 0;
+        let result = OakStatus::from_i32(unsafe {
+            oak_abi::handle_clone(handle, &mut cloned_handle as *mut Handle) as i32
+        })
+        .expect("handle_clone returned invalid oak status");
+        if result != OakStatus::Ok {
+            panic!("Failed to clone handle: {:?}", result);
+        }
+        cloned_handle
+    }
+
+    fn drop_handle(handle: Handle) {
+        // The channel may have already been closed, or it may be invalid due to extracting it for
```

Comment:
I think we should check for invalid handle here (and also in clone) and deal with that case explicitly, whether it is about ignoring it, or producing a more precise error message.

Reply:
Sure, maybe we can ignore `INVALID_HANDLE` in `drop`, and `panic!` with a more helpful message in `clone`?